extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,321 @@
+        try {
+            currentFragment = currFragmentFuture.get(readTimeout.toNanos(), TimeUnit.NANOSECONDS);
+        } catch (final InterruptedException | ExecutionException | TimeoutException | CancellationException e) {
+            final String operation =
+                    "fetching fragment " + currFragmentIndex + " for file " + key + " in S3 bucket " + bucket;
+            throw handleS3Exception(e, operation);
+        }
+
+        // Copy the bytes from fragment from the offset up to the min of remaining fragment and destination bytes.
+        // Therefore, the number of bytes read by this method can be less than the number of bytes remaining in the
+        // destination buffer.
+        final int fragmentOffset = (int) (localPosition - (currFragmentIndex * fragmentSize));
+        currentFragment.position(fragmentOffset);
+        final int limit = Math.min(currentFragment.remaining(), destination.remaining());
+        final int originalBufferLimit = currentFragment.limit();
+        currentFragment.limit(currentFragment.position() + limit);
+        destination.put(currentFragment);
+        // Need to reset buffer limit, so we can read from the same buffer again in future
+        currentFragment.limit(originalBufferLimit);
+        position = localPosition + limit;
+        return limit;
+    }
+
+    private int fragmentIndexForByteNumber(final long byteNumber) {
+        return Math.toIntExact(byteNumber / fragmentSize);
+    }
+
+    /**
+     * @return A {@link CompletableFuture} that will be completed with the bytes of the fragment
+     */
+    private CompletableFuture<ByteBuffer> sendAsyncRequest(final int fragmentIndex) {
+        final long readFrom = (long) fragmentIndex * fragmentSize;
+        final long readTo = Math.min(readFrom + fragmentSize, size) - 1;
+        final String range = "bytes=" + readFrom + "-" + readTo;
+        return s3AsyncClient.getObject(builder -> builder.bucket(bucket).key(key).range(range),
+                new ByteBufferAsyncResponseTransformer<>(fragmentSize));
+    }
+
+    private UncheckedDeephavenException handleS3Exception(final Exception e, final String operationDescription) {
+        if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+            return new UncheckedDeephavenException("Thread interrupted while " + operationDescription, e);
+        } else if (e instanceof ExecutionException) {
+            return new UncheckedDeephavenException("Execution exception occurred while " + operationDescription, e);
+        } else if (e instanceof TimeoutException) {
+            return new UncheckedDeephavenException("Operation timeout while " + operationDescription + " after waiting "
+                    + "for duration " + readTimeout, e);
+        } else if (e instanceof CancellationException) {
+            return new UncheckedDeephavenException("Cancelled an operation while " + operationDescription, e);
+        }
```

Comment:
Optional, stylistic suggestion only:
```suggestion
        if (e instanceof InterruptedException) {
            Thread.currentThread().interrupt();
            return new UncheckedDeephavenException("Thread interrupted while " + operationDescription, e);
        }
        if (e instanceof ExecutionException) {
            return new UncheckedDeephavenException("Execution exception occurred while " + operationDescription, e);
        }
        if (e instanceof TimeoutException) {
            return new UncheckedDeephavenException(String.format(
                    "Operation timeout while %s after waiting for duration %s", operationDescription, readTimeout), e);
        }
        if (e instanceof CancellationException) {
            return new UncheckedDeephavenException("Cancelled an operation while " + operationDescription, e);
        }
```